### PR TITLE
move to ESM and what not

### LINKED ISSUE #1288: ESM (ES6 module) distribution
**Describe the feature**
Publish standard JS modules to npm, instead of UMD.

**Why is this feature necessary?**
All major browsers can load standard JS modules now, so importing this library becomes much easier.

### PR DESCRIPTION
**Pull request by 43081j**
👋 

I had this lying around and decided to finish it off recently in my fork as it was needed for trying out in some web component stuff I had setup.

Fixes #1288

Is this something you'd be interested in? Happy to keep it in my fork if not and close this PR.

---

Here's a summary of what happened in this branch:

* Sources separated into modules
* All sources are valid ESM (`src/*`)
* Moved from prototypical concepts to classes
* Introduced babel to transpile to CJS modules (`lib/*`) for when node users require this
* Introduced rollup to bundle to a browser-compatible library (`marked.min.js`) which behaves the same as the original
* Introduced `files` to package.json so `npm pack` and publish will package only bin/lib/man (CJS sources)
* Introduced a `module` entry to package.json so bundlers can resolve the ESM sources
* Replaced all unscoped vars and what not to `let`/`const`

All tests pass, build succeeds.

Have not been using it in a browser myself so it seems that would need further testing.

Whats missing:

* The browser bundle isn't minified (something like `babel-preset-minify` could achieve that)
* Performance comparison
* Build will fail on CI as `npm run build` needs adding to travis.yml

---

Again, if this is of any use to you just let me know and ill get back on it.

cc @joshbruce

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -1,3 +0,0 @@
-.git*
-test/
-docs
```
Comment (styfle) [inline .npmignore]: Why was .npmignore removed?

File: package.json [hunk 2]
```diff
@@ -51,11 +60,15 @@
...
   },
   "engines": {
     "node": ">=0.10.0"
-  }
+  },
+  "files": [
+    "bin/**",
+    "lib/**",
+    "man/**",
+    "README.md"
```
Comment (styfle) [inline package.json]: I see now, this is a better solution than npmignore

File: .npmignore (hunk 1 above)
Comment (43081j) [reply .npmignore]: Yup the recommended method of defining what gets packaged is the `files` array now and thngs like `node_modules` are automatically excluded.

File: package.json (hunk 2 above)
Comment (43081j) [reply package.json]: Note to self: this needs `src/**` adding for compatibility with browsers importing the ESM source, though.

File: package.json [hunk 3]
```diff
@@ -51,11 +60,15 @@
     "test:lint": "eslint bin/marked .",
     "test:redos": "eslint --plugin vuln-regex-detector --rule '\"vuln-regex-detector/no-vuln-regex\": 2' lib/marked.js",
     "bench": "node test --bench",
-    "lint": "eslint --fix bin/marked .",
-    "build": "uglifyjs lib/marked.js -cm  --comments /Copyright/ -o marked.min.js",
```
[the remark this answers is not in the document]
Comment (43081j) [reply package.json]: the rollup call produces it, unminified at the min but introducing `babel-preset-minify` should sort that.

**Comment by styfle** [comment]
The [tests](https://travis-ci.org/markedjs/marked/builds/436376656) are failing with Error: Cannot find module './lexer'

**Comment by 43081j** [comment]
Yup this is because the travis config needs updating to `npm run build` beforehand.

**Comment by styfle** [comment]
@43081j You should be able to modify `.travis.yml` and add a build step

**Comment by 43081j** [comment]
Yup its part of the list in my original post of whats left to do. simple enough, just didn't get around to it

**Comment by styfle** [comment]
Excellent! Let's wait to see what the rest of the team thinks.

File: .travis.yml [hunk 4]
```diff
@@ -7,13 +7,12 @@ jobs:
 
   include:
     - stage: unit tests 👩🏽‍💻
-      script: npm run test:unit
+      script: npm run build && npm run test:unit
```
Comment (UziTech) [inline .travis.yml]: since `npm test` already includes `npm run build` it is redundant here and below

File: package.json [hunk 5]
```diff
@@ -23,39 +24,53 @@
...
-    "test:redos": "eslint --plugin vuln-regex-detector --rule '\"vuln-regex-detector/no-vuln-regex\": 2' lib/marked.js",
+    "test:lint": "eslint bin/marked \"test/**/*.js\"",
+    "test:redos": "eslint --plugin vuln-regex-detector --rule '\"vuln-regex-detector/no-vuln-regex\": 2' src/marked.js",
     "bench": "node test --bench",
-    "lint": "eslint --fix bin/marked .",
-    "build": "uglifyjs lib/marked.js -cm  --comments /Copyright/ -o marked.min.js",
     "preversion": "npm run build && (git diff --quiet || git commit -am 'minify')"
   },
   "engines": {
     "node": ">=0.10.0"
```
Comment (UziTech) [inline package.json]: should this be updated to `">=6.0.0"`?

**Comment by UziTech** [comment]
Do we need to add [browserslist](https://babeljs.io/docs/en/babel-preset-env#browserslist-integration) to `package.json` for `babel-preset-env`?

**Comment by 43081j** [comment]
We probably should but i wasn't too sure what the correct targets should be.

The ESM sources account for any browsers supporting ES modules.

The CJS sources should only be used by node, so our babel-preset config should probably be `node: true`.

The bundle should be consumed by non-ESM browsers so its actually a different target to `lib/` I guess (overridable via rollup.config.js if we must). We would have to find the minimum browser version we could support.

File: package.json [hunk 6]
```diff
@@ -23,39 +24,53 @@
...
     "test:specs": "npm test -- test/specs/**/*-spec.js",
     "test:cm": "npm test -- test/specs/commonmark/**/*-spec.js",
     "test:gfm": "npm test -- test/specs/gfm/**/*-spec.js",
     "test:marked": "npm test -- test/specs/marked/**/*-spec.js",
     "test:old": "node test",
-    "test:lint": "eslint bin/marked .",
-    "test:redos": "eslint --plugin vuln-regex-detector --rule '\"vuln-regex-detector/no-vuln-regex\": 2' lib/marked.js",
+    "test:lint": "eslint bin/marked \"test/**/*.js\"",
+    "test:redos": "eslint --plugin vuln-regex-detector --rule '\"vuln-regex-detector/no-vuln-regex\": 2' src/marked.js",
     "bench": "node test --bench",
```
Comment (UziTech) [inline package.json]: `"test:old"` and `"bench"` will need to be prepended with `npm run build &&`

File: rollup.config.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+export default {
+  input: 'src/marked.js',
+  output: {
+    file: 'marked.min.js',
+    format: 'umd',
+    name: 'marked'
+  },
+  plugins: [babel()]
+}
```
Comment (UziTech) [inline rollup.config.js]: Missing semicolon (sorry but my IDE nags me 🙃)

**Comment by UziTech** [comment]
benchmarks seem to be consistent 👍:

```
// master
marked completed in 5809ms.
marked (gfm) completed in 7043ms.
marked (pedantic) completed in 5590ms.
commonmark completed in 8877ms.
markdown-it completed in 6774ms.
markdown.js completed in 22230ms.

// this pr
marked completed in 6027ms.
marked (gfm) completed in 6981ms.
marked (pedantic) completed in 5792ms.
commonmark completed in 9057ms.
markdown-it completed in 8349ms.
markdown.js completed in 20882ms.
```

**Comment by 43081j** [comment]
What sort of time frame would you be looking at for 1.x?

To be honest, ES modules are still a bit flaky with node which is why we're forced to have 3 builds (ESM, CJS, browser).

I think if we try wait for node to have a better implementation, we'll be here _a while_.

It may be worth us looking into using this:
https://github.com/standard-things/esm

instead of shipping two copies of the sources 🤔  but its a fairly big dependency to be shipping i guess

**Comment by styfle** [comment]
I looked this over again and I am becoming more reserved the more I look at it.

> ES modules are still a bit flaky with node which is why we're forced to have 3 builds (ESM, CJS, browser).

This PR is adding more complexity to a relatively small project and I'm not sure the value added here is worth it at this time.

Let's do a pros/cons list

## Pros

1. Developers get to use multiple files - maybe avoid merge conflicts?
2. Developers get to use ES6 syntax - less writing and more readable in most cases
3. (maybe) Users can use ESM sources with bundlers to maybe treeshake output

## Cons

1. The ESM source will not run in the browser as-is (due to bare specifiers)
2. The ESM source will not run in Node.js as-is (due to .mjs and experimental)
3. The bundle size is actually larger by a few KB (was 42 KB and now 47 KB)
4. We'll be publishing more sources files so removing those sources is a breaking change (probably can do this in 1.x)
5. More complexity due to more build tools requires maintenance when upgrading versions: babel, rollup, and several more plugins
6. (maybe) Some js user's don't know ES6 syntax and won't be able to contribute
7. (maybe) Build time will increase slightly due to more build targets

**Comment by joshbruce** [comment]
@styfle brings up some very good points.

Tagging #746 as it describes a method for modularizing the codebase with straight ES5 or 6 (??) (doing this off the top of my head).

To @43081j's question, the core team is really trying to get the community to step up more. One of the issues of the past (before the current core team) was an expectation became set that the core team would do everything...so I cannot tell you how awesome it is to have you post this PR, whether it's accepted or not...anyhoo, the core team primarily sets direction, creates tools, and ensures consistency of the codebase and approach; so, all that to say, timeframe is up to the broader community as the core team will focus more on helping and coordination than doing (though we do some of that when we can). A refrain that became popular for a minute was "1.0 could happen tomorrow or may never happen, it's up to the community" (or something like that).

See https://github.com/markedjs/marked/issues/1106 for the details and drama...in fact, probably time to write another one.

**Throwing copper:** I think, based on @styfle's thoughts, we should close this for now - reference to what I think is the main modularization issue - and maybe resurrect at that time. Spec-compliance and keeping the infrastructure clean per the roadmap and milestones (https://github.com/markedjs/marked/milestones) unless we find something that can minimize impact on legacy users or out right fake backward compatibility. Not to say the new stuff isn't nice, just may not be ready to be part of main..

**Comment by 43081j** [comment]
This is meant to introduce the ability to use the ES modules in browsers _as is_. So if that means I forgot to (and need to) add the file extension to each import, I should.

The bundle size is _of course_ larger, it isn't even minified yet (and rollup will tree-shake, which it hasn't done yet either).

The rest of @styfle's points are fairly small IMHO, tooling being introduced isn't a problem, nor is build time (it won't be noticeable most likely). Git history shouldn't be an issue either as the files should be detected as being moved.

**Comment by Martii** [comment]
> Developers get to use ES6 syntax - less writing and more readable in most cases

This is actually a con in our book. ES6 may be supported however it's *very subjective on readability*. Examples written in ES5 are a good base for projects that don't use ES6+ and extending it to the more quirky ES6+ syntaxes.

**Pull request closed without merging**